text/0088-tvm-update.md

```diff
@@ -0,0 +1,142 @@
+- **replaces**: -
+- **replaced by**: -
+
+# Summary
+
+This proposal suggests to
+- add new op-codes to TVM related to block data retrieving, hashing, cryptographic signing and message sending
+- include into `c7` register of TVM smart-contract' own code, additional data on transaction and information about masterchain blocks
+ 
+# Motivation
+
+Current TVM implementation makes it hard to implement some common thing. Proposal aims to make it easier and cheaper to:
+- hash big chanks of data (including non-sha256 hashes)
+- work with Bitcoin/Ethereum compatible signtatures
+- prove TON onchain data
+- calculate fee for sending messages
+
+# Guide
+
+Draft implementation is available [here](https://github.com/SpyCheese/ton/commits/new-opcodes).
+
+This proposal suggests to 
+1. Extend **c7** tuple from 10 to 14 elements:
+   - **10**: code of the smart contract.
+   - **11**: value of the incoming message.
+   - **12**: fees collected in the storage phase.
+   - **13**: information about previous blocks.
+2. Add the following TVM op-codes:
+   - **c7** primitives: `MYCODE`, `INCOMINGVALUE`, `STORAGEFEES`, `PREVBLOCKSINFOTUPLE`
+   - Block primitives: `PREVMCBLOCKS`, `PREVKEYBLOCK`
+   - Hash primitives: `HASHSTART`, `HASHEND`, `HASHENDST`, `HASHINFO`, `HASHAPPU`, `HASHAPPI`, `HASHAPPS`, `HASHAPPB`
+   - Cryptography primitives: `ECRECOVER`
+   - Message primitives: `SENDMSG`
+
+3. Change sending fee calculation: deduct fee related to message estimation even in the case of failed sending.
+
+# Specification
+
+## c7 update
+**10th** element of **c7** tuple contains cell with executed smart-contract code (from init_state of incoming message if applicable)
+
+**11th** element of **c7** tuple contains value of incoming message. It is a tuple with two elements: incoming value (same value which is put on stack as 4th element) and a dictionary of extra currencies. If not applicable, the value is `[0, null]`.
+
+**12th** element of **c7** tuple contains fee debited in storage phase.
+
+**13th** elements of **c7** contains of tuple with two elements: `last_mc_blocks` and `prev_key_block`.
+- `last_mc_blocks` is a tuple of up to 16 elements which contains previous most recent masterchain blocks info (ordered by seqno). Each block is represented as tuple of 5 integers: `(workchain, shard, seqno, root_hash, file_hash)`
+- `prev_key_block` is a tuple of 5 integers `(workchain, shard, seqno, root_hash, file_hash)` corresponding to most recent key block.
+
+## TVM update
+### New stack type
+This proposal suggests to introduce new type in TVM `Hasher`: it is opaque type used to consequently hash data chunks. This type behave the same way as others and can be considered immutable: `HASHAPP*` ops (described below) "consume" old hasher and return updated one (that way if old hasher was copied, it's copies are completely independent). This object is not serializable and thus can not be returned from TVM (for instance as result of get-method).
```

Comment:
I think it's better to use the name "Digest" as in popular languages like Java or GoLang, because this is a more abstract name for hashing/checksum calculation algorithms.

Reply:
It looks like "digest" refers to result of hashing, here we use `Hasher` for intermediate object with some internal state.

Reply:
Isn't this state could be implemented via a generic slice?

Reply:
Well, any state can be can be represented as some state. The question do we really need it?

Reply:
4304417

Reply:
UQAP7O1Pm6kHiSerD6L52UIXPYyeYId2ha5WsHWPwOCx_-3p